File: Week_01/239/sliding_window_maximum.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = deque()
        res = []
        for i, n in enumerate(nums):
            while dq and dq[0] <= i - k:
                dq.popleft()
            while dq and nums[dq[-1]] < n:
                dq.pop()
            dq.append(i)
            if i >= k - 1:
                res.append(nums[dq[0]])
        return res

    def test(self):
        for nums, k, target in [
            ([], 3, []),
            ([1, 2], 1, [1, 2]),
            ([1, 3, -1, -3, 5, 3, 6, 7], 3, [3, 3, 5, 5, 6, 7]),
        ]:
            ans = self.maxSlidingWindow(nums, k)
            assert ans == target, f"target: {target}, ans: {ans}"
        print("all done")


if __name__ == "__main__":
    Solution().test()

File: Week_01/239/sliding_window_maximum.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	var res, windows []int

	for i, n := range nums {
		if len(windows) > 0 && windows[0] <= i-k {
			windows = windows[1:]
		}

		for len(windows) > 0 && nums[windows[len(windows)-1]] < n {
			windows = windows[:len(windows)-1]
		}
		windows = append(windows, i)

		if i >= k-1 {
			res = append(res, nums[windows[0]])
		}
	}
	return res
}
